grails-app/services/org/pih/warehouse/inventory/RecordStockProductInventoryTransactionService.groovy

```diff
@@ -0,0 +1,105 @@
+package org.pih.warehouse.inventory
+
+import org.pih.warehouse.api.AvailableItem
+import org.pih.warehouse.core.Constants
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.product.Product
+
+import grails.gorm.transactions.Transactional
+
+import javax.xml.bind.ValidationException
+
+@Transactional
+class RecordStockProductInventoryTransactionService {
+
+    InventoryService inventoryService
+    TransactionIdentifierService transactionIdentifierService
+
+    Transaction createInventoryBaselineTransaction(
+            Location facility,
+            Date transactionDate = new Date(),
+            List<AvailableItem> items
+    ) {
+        TransactionType transactionType = TransactionType.read(Constants.INVENTORY_BASELINE_TRANSACTION_TYPE_ID)
+
+        Transaction transaction = new Transaction(
+                source: facility,
+                inventory: facility.inventory,
+                transactionDate: transactionDate,
+                transactionType: transactionType,
+        )
+
+        transaction.transactionNumber = transactionIdentifierService.generate(transaction)
+
+        for (AvailableItem availableItem : items) {
+            TransactionEntry transactionEntry = new TransactionEntry(
+                    quantity: availableItem.quantityOnHand,
+                    binLocation: availableItem.binLocation,
+                    inventoryItem: availableItem.inventoryItem,
+            )
+
+            transaction.addToTransactionEntries(transactionEntry)
+        }
+
+        if (!transaction.save()) {
+            throw new ValidationException("Invalid transaction", transaction.errors)
+        }
+        return transaction
+    }
+
+    List<AvailableItem> getAvailableItems(
+            Inventory inventory,
+            Product product
+    ) {
+        List<AvailableItem> availableItems = new ArrayList<>()
+        List<TransactionEntry> transactionEntries = inventoryService.getTransactionEntriesByInventoryAndProduct(inventory, [product])
+        List<BinLocationItem> binLocationItems = inventoryService.getQuantityByBinLocation(transactionEntries)
+
+        for (BinLocationItem binLocationItem : binLocationItems) {
+            AvailableItem availableItem = new AvailableItem(
+                    quantityAvailable: binLocationItem.quantity,
+                    binLocation: binLocationItem.binLocation,
+                    inventoryItem: binLocationItem.inventoryItem,
+            )
+            availableItems.add(availableItem)
+        }
+
+        return availableItems
+    }
+
+    Integer getTransactionEntryQuantity(List<AvailableItem> availableItems, Integer newQuantity, InventoryItem inventoryItem) {
+        AvailableItem availableItem = availableItems.find {
```

Comment:
I implemented a quite similar solution for inventory import except I made this a method in the ProductAvailabilityService and instead of returning a list of AvailableItem, I return a Map<String, AvailableItem> keyed on [product + bin + lot] for easy lookups. I'm going to make a PR for that soon, but hopefully we can unify the behaviour